llvm/include/llvm/Bitstream/BitstreamWriter.h

```diff
@@ -415,6 +425,113 @@ class BitstreamWriter {
+        if (Vals[RecordIdx] != Op.getLiteralValue())
+          return false;
+        ++RecordIdx;
+      } else if (Op.getEncoding() == BitCodeAbbrevOp::Array) {
+        // Arrays store an arbitrary number of values with a specified encoding.
+        // Check the value of each element of the array to see if it works.
+        assert(i + 2 == e && "array op not second to last?");
+        const BitCodeAbbrevOp &EltEnc = Abbv->getOperandInfo(++i);
+        unsigned MaxBitwidth = getMaxBitWidth(EltEnc);
+
+        if (Blob.data()) {
+          // 'Blob' takes the place of entries in the Vals array. There should
+          // be no more Vals.
+          if (RecordIdx != Vals.size())
+            return false;
+          // If the bitwidth of the encoding is below 8, we must validate the
+          // blob values.
+          if (MaxBitwidth < 8) {
+            if (!llvm::all_of(
+                    Blob, [&](unsigned B) { return isUIntN(MaxBitwidth, B); }))
+              return false;
+          }
+        } else {
+          // Check each Val
+          for (uintty e = Vals.size(); RecordIdx != e; ++RecordIdx)
+            if (!isUIntN(MaxBitwidth, Vals[RecordIdx]))
+              return false;
+        }
+        return true; // Checked all data!
+      } else if (Op.getEncoding() == BitCodeAbbrevOp::Blob) {
+        // Blobs store an arbitrary number of 8-bit values
+        assert(i + 1 == e && "blob op not last?");
+
+        if (Blob.data()) {
+          // 'Blob' takes the place of entries in the Vals array. There should
+          // be no more Vals.
+          if (RecordIdx != Vals.size())
+            return false;
+          // We already know the encoding fits, so don't need to check.
+        } else {
+          if (!llvm::all_of(Vals.slice(RecordIdx),
+                            [](uintty B) { return isUInt<8>(B); }))
+            return false;
+        }
+        return true; // Checked all data!
+      } else {
+        // Check if the specified value fits the encoding.
+        if (RecordIdx >= Vals.size())
+          return false;
+
+        unsigned MaxBitwidth = getMaxBitWidth(Op);
+
+        if (!isUIntN(MaxBitwidth, Vals[RecordIdx]))
+          return false;
+        ++RecordIdx;
+      }
+    }
+
+    // Finally, check that we used all the input data.
+    return !Blob.data() && RecordIdx == Vals.size();
+  }
+
+  /// Search for a valid abbrev which can encode the provided data.
+  /// Return it if found, otherwise return 0.
+  template <typename uintty>
+  unsigned FindValidAbbrev(unsigned Code, ArrayRef<uintty> Vals,
+                           StringRef Blob) {
+    if (auto it = CodeAbbrevIndex.find(Code); it != CodeAbbrevIndex.end()) {
+      for (unsigned AbbrevToTry : it->second) {
+        const BitCodeAbbrev *Abbv =
+            CurAbbrevs[AbbrevToTry - bitc::FIRST_APPLICATION_ABBREV].get();
+        if (CheckAbbrevValidity(Vals, Blob, Abbv))
+          return AbbrevToTry;
+      }
+    }
```

Comment:
I think that if we went with the tablegen route, then we would be generating code similar to the existing manual approach, for performance reasons.

I am assuming this passes the google monorepo build without any significant CPU time performance regressions, is that right?